### PR TITLE
skipper: refactor test options

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
Remove unexported test options to allow using them outside of skipper package.

* refactor testOptions into Options fields
* add shutdown channels

- [x] requires https://github.com/zalando/skipper/pull/2418

### CODE REVIEW

**Comment by AlexanderYastrebov** [comment]
This would be used by https://github.com/zalando/skipper/pull/2416

File: skipper_test.go [hunk 1]
```diff
@@ -1043,4 +1041,5 @@ func TestDataClients(t *testing.T) {
 		t.Fatalf("Failed to GET the status of routes file route: %d", rsp.StatusCode)
 	}
 
+	close(o.Shutdown)
```
[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply skipper_test.go:1044]: It fails anyway and even on master
```
$ go test . -count=10 -run=TestDataClients -v
=== RUN   TestDataClients
time="2023-06-21T19:14:29+02:00" level=info msg="route settings, reset, upsert count: 1"
time="2023-06-21T19:14:29+02:00" level=info msg="route settings, reset, delete count: 0"
time="2023-06-21T19:14:29+02:00" level=info msg="route settings, reset, upsert count: 1"
time="2023-06-21T19:14:29+02:00" level=info msg="route settings, reset, delete count: 0"
time="2023-06-21T19:14:29+02:00" level=info msg="route settings received"
time="2023-06-21T19:14:29+02:00" level=info msg="route settings received"
time="2023-06-21T19:14:29+02:00" level=info msg="route settings applied"
time="2023-06-21T19:14:29+02:00" level=info msg="route settings applied"
--- PASS: TestDataClients (0.11s)
=== RUN   TestDataClients
--- PASS: TestDataClients (0.01s)
=== RUN   TestDataClients
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x17706b9]

goroutine 150 [running]:
github.com/zalando/skipper/swarm.(*Swarm).Leave(0x0)
        /home/ayastrebov/projects/go/src/github.com/zalando/skipper/swarm/swarm.go:407 +0x19
github.com/zalando/skipper.run({0x0, {0xc00013a060, 0x2, 0x2}, {0x0, 0x0, 0x0}, {0x1bb9414, 0x5}, {0x0, ...}, ...}, ...)
        /home/ayastrebov/projects/go/src/github.com/zalando/skipper/skipper.go:1985 +0x5dda
created by github.com/zalando/skipper.TestDataClients
        /home/ayastrebov/projects/go/src/github.com/zalando/skipper/skipper_test.go:1027 +0xafb
FAIL    github.com/zalando/skipper      0.136s
FAIL
```

[the remark this answers is not in the document]
Comment (szuecs) [reply skipper_test.go:1044]: I don't like exposed channels for shutdown.

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply skipper_test.go:1044]: Its ok to not like something. If you explain why maybe we can find a better way :)

File: skipper.go [hunk 2]
```diff
@@ -889,7 +887,12 @@ type Options struct {
 	// filters.
 	LuaSources []string
 
-	testOptions
+	// Shutdown is used to start graceful shutdown of Skipper by closing it.
+	// If not set Skipper will start graceful shutdown after receiving SIGTERM.
+	Shutdown chan struct{}
```
Comment (szuecs) [inline skipper.go:892]: Not a good idea imo to expose channels

File: skipper_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (szuecs) [reply skipper_test.go:1044]: A channel is not a good api. If you expose it like that then everyone has to use it as library user. It should be internal and I don't see why you need to expose it.

**Review by szuecs — CHANGES_REQUESTED** [review]
Do not expose channels in skipper library, because it is not a good api for library users and I don't see why we would need it.

**Comment by szuecs** [comment]
We discussed that you can hopefully use export_test.go to expose an API to pass channels, if routesrv  does it similar we can not use this to start both and then it's fine to pass it via Options.

File: skipper_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply skipper_test.go:1044]: This change does not expose any internal channels but allows user to specify optional channels to trigger and signal shutdown.

**Comment by AlexanderYastrebov** [comment]
>  if routesrv does it similar

routesrv also does not support programmatic shutdown atm

**Comment by AlexanderYastrebov** [comment]
We need to think more about programmatic shutdown for both skipper and routesrv.
Besides adding channels to options we may add exported RunWithShutdown functions.

Closing this for now in favor of #2419

**Pull request closed without merging**